app/gui/view/src/project.rs

```diff
@@ -376,23 +402,55 @@ impl View {
         let input_change_delay = frp::io::timer::Timeout::new(network);
         let searcher_open_delay = frp::io::timer::Timeout::new(network);
 
-        if let Some(window_control_buttons) = &*model.window_control_buttons {
-            let initial_size = &window_control_buttons.size.value();
-            model.graph_editor.input.space_for_window_buttons(initial_size);
-            frp::extend! { network
-                graph.space_for_window_buttons <+ window_control_buttons.size;
-                eval_ window_control_buttons.close      (model.on_close_clicked());
-                eval_ window_control_buttons.fullscreen (model.on_fullscreen_clicked());
-            }
+        frp::extend! { network
+            init <- source::<()>();
         }
 
-        let shape = scene.shape().clone_ref();
 
+        // === Top Bar ===
+
+        let window_control_buttons_width =
+            if let Some(window_control_buttons) = &*model.window_control_buttons {
+                frp::extend! { network
+                    window_control_buttons_size <- all(init, window_control_buttons.size)._1();
+                    window_control_buttons_width <- window_control_buttons_size.map(|size| size.x);
+                    eval_ window_control_buttons.close (model.on_close_clicked());
+                    eval_ window_control_buttons.fullscreen (model.on_fullscreen_clicked());
+                }
+                window_control_buttons_width
+            } else {
+                frp::extend! { network
+                    window_control_buttons_width <- source::<f32>();
+                    window_control_buttons_width_stream <- window_control_buttons_width.identity();
+                }
+                window_control_buttons_width.emit(0.0);
+                window_control_buttons_width_stream
+            };
+        let scene_shape = scene.shape().clone_ref();
+        let dashboard_button = &model.dashboard_button.frp;
         frp::extend! { network
-            init <- source::<()>();
-            shape <- all(shape, init)._0();
-            eval shape ((shape) model.on_dom_shape_changed(shape));
+            frp.source.dashboard_button_pressed <+ dashboard_button.pressed;
+            dashboard_button_size <- all(init, dashboard_button.size)._1();
+            dashboard_button_width <- dashboard_button_size.map(|size| size.x);
+            top_bar_update <- all(
+                init,
+                scene_shape,
+                window_control_buttons_width,
+                dashboard_button_width
+            );
+            top_bar_width <- top_bar_update.map(
+                f!([model] ((_, scene_shape, window_control_buttons_width, dashboard_button_width))
```

Comment:
nit: you can skip `[model]` if you have an expression like `model.bla` (like in this case)

Reply:
How does it work? When is it needed then?

Reply:
So `f!` macro has a special case for expressions like `foo.bar`. Such expressions are equivalent to `[foo] foo.bar()`. In other cases you need to manually `clone_ref` the `foo`, by using square brackets: `[foo] if bla { foo.bar() }`

Reply:
Ah, gotcha! Thanks for explaining!